rust/datafusion/src/sql/planner.rs

```diff
@@ -523,10 +523,14 @@ impl<S: SchemaProvider> SqlToRel<S> {
 
                             let mut safe_args: Vec<Expr> = vec![];
                             for i in 0..rex_args.len() {
-                                safe_args.push(
-                                    rex_args[i]
-                                        .cast_to(fm.args()[i].data_type(), schema)?,
-                                );
+                                let expr = if fm.args()[i]
+                                    .contains(&rex_args[i].get_type(schema)?)
+                                {
+                                    rex_args[i].clone()
+                                } else {
+                                    rex_args[i].cast_to(&fm.args()[i][0], schema)?
```

Comment:
Adding a test for a user defined function that takes no args would probably be good -- an example of such a function might be `rand()`

Reply:
I am not sure we can easily declare a zero-argument UDF: we do not know how many rows we should build because we have no `ArrayRef` to count them. IMO this is something that we will have to re-visit.

I will add a test for 2 arguments of various types.

Reply:
makes sense -- I guess I would suggest we add a test to ensure the system behaves reasonably (aka generates an error rather than a crash) with 0 argument UDFs -- but if there is something challenging about implementing them it doesn't have to be done in this PR